src/main/java/com/itextpdf/rups/io/PdfObjectTreeEdit.java

```diff
@@ -0,0 +1,368 @@
+    of this program must display Appropriate Legal Notices, as required under
+    Section 5 of the GNU Affero General Public License.
+
+    In accordance with Section 7(b) of the GNU Affero General Public License,
+    a covered work must retain the producer line in every PDF that is created
+    or manipulated using iText.
+
+    You can be released from the requirements of the license by purchasing
+    a commercial license. Buying such a license is mandatory as soon as you
+    develop commercial activities involving the iText software without
+    disclosing the source code of your own applications.
+    These activities include: offering paid services to customers as an ASP,
+    serving PDFs on the fly in a web application, shipping iText with a closed
+    source product.
+
+    For more information, please contact iText Software Corp. at this
+    address: sales@itextpdf.com
+ */
+package com.itextpdf.rups.io;
+
+import com.itextpdf.kernel.pdf.PdfName;
+import com.itextpdf.rups.controller.PdfReaderController;
+import com.itextpdf.rups.event.NodeAddArrayChildEvent;
+import com.itextpdf.rups.event.NodeAddDictChildEvent;
+import com.itextpdf.rups.event.NodeDeleteArrayChildEvent;
+import com.itextpdf.rups.event.NodeDeleteDictChildEvent;
+import com.itextpdf.rups.event.NodeUpdateArrayChildEvent;
+import com.itextpdf.rups.event.NodeUpdateDictChildEvent;
+import com.itextpdf.rups.event.RupsEvent;
+import com.itextpdf.rups.model.ErrorDialogPane;
+import com.itextpdf.rups.view.itext.PdfObjectPanel;
+import com.itextpdf.rups.view.itext.treenodes.PdfObjectTreeNode;
+
+import javax.swing.undo.CannotRedoException;
+import javax.swing.undo.CannotUndoException;
+import javax.swing.undo.UndoableEdit;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Stack;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class PdfObjectTreeEdit implements UndoableEdit {
+
+    enum ObjectType {
+        ARRAY, DICT
+    }
+    enum ActionType {
+        ADD, DELETE, UPDATE
+    }
+
+    ObjectType objectType;
+    ActionType actionType;
+    PdfReaderController controller;
+    int editType; // The type of edit, as defined by the static bytes within RupsEvent.java (Refactor to Enum?)
+    PdfObjectTreeNode parent; // Parent Node
+    PdfObjectTreeNode oldChild; // Child Node
+    PdfObjectTreeNode child; // Child Node
+    int index; // Index at which the edit occurred.
+    PdfName key; // Index at which the edit occurred.
+    List<UndoableEdit> addedEdits;
+    boolean dead;
+
+    public PdfObjectTreeEdit(PdfReaderController controller, RupsEvent editEvent) {
+        this.controller = controller;
+        this.addedEdits = new LinkedList<>();
+        this.dead = false;
+        extractEditData(editEvent);
+    }
+
+    private void extractEditData(RupsEvent editEvent) {
+        // TODO: Maybe we can add some granularity to these events?
```

Comment:
@michaeldemey is this ToDo worth keeping in for future brainstorming or best removed for cleanliless?